src/olympia/addons/views.py

```diff
@@ -982,4 +983,10 @@ def filter_queryset(self, qs):
         taar_enable = self.request.GET.get('recommended', '').lower() == 'true'
         guids, self.ab_outcome, self.fallback_reason = (
             get_addon_recommendations(guid_param, taar_enable))
-        return qs.query(query.Bool(must=[Q('terms', guid=guids)]))
+        results_qs = qs.query(query.Bool(must=[Q('terms', guid=guids)]))
+
+        if results_qs.count() != 4 and is_outcome_recommended(self.ab_outcome):
```

Comment:
Don't use `.count()`, because that will create another ES query, which can be avoided if all add-ons are present. Instead you want to call `.execute()` on the queryset, storing this into a variable, and check `.hits.total` on it.

Add a test proving that this does only one ES query in the best case scenario.